source/common/router/upstream_request.cc

```diff
@@ -93,6 +93,28 @@ UpstreamRequest::~UpstreamRequest() {
         FilterUtility::percentageOfTimeout(response_time, parent_.timeout().per_try_timeout_));
   }
 
+  // Ditto for request/response size histograms.
+  Upstream::ClusterRequestResponseSizeStatsOptRef req_resp_stats_opt =
+      parent_.cluster()->requestResponseSizeStats();
+  if (req_resp_stats_opt.has_value()) {
+    auto& req_resp_stats = req_resp_stats_opt->get();
+    req_resp_stats.upstream_rq_headers_size_.recordValue(parent_.downstreamHeaders()->byteSize());
+
+    auto rq_body_size = stream_info_.bytesSent();
+    if (rq_body_size != 0) {
+      req_resp_stats.upstream_rq_body_size_.recordValue(rq_body_size);
+    }
+
+    if (response_headers_size_.has_value()) {
+      req_resp_stats.upstream_rs_headers_size_.recordValue(response_headers_size_.value());
+
+      auto rs_body_size = stream_info_.bytesSent();
+      if (rs_body_size != 0) {
```

Comment:
So unlike response headers, here I _think_ we want to either
1) differentiate between 0 length bodies and no bodies or
2) always record 0 if we can't differentiate

Previously we did add to the histogram for empty bodies and I think not doing that may be a functional change some folks don't want.

Reply:
Hmm, I don't think anyone would complain given the stats aren't currently generated :p 

The way I see it is – and really this is how we are using this internally – that we use this to track regressions in payload sizes (oh you did what to that Javascript lib again). So skipping 0 length bodies is probably alright, if what you are tracking really is that your bodies don't get bloated and not so much how many empty bodies were sent.

Reply:
Wait not at all?  yikes, it looks from code like it'd be generated, just not for retries.

Regardless I _think_ we should track 0 length bodies, because while they're not needed for your use case, they could be for others' 
If you actively prefer not doing so, I think you need to update docs to make that explicit, and then if someone wants it later thy can add a knob, but I'd prefer just tracking all bodies.

Reply:
Hmm, in the case of the request's body we have:

https://github.com/envoyproxy/envoy/blob/main/envoy/stream_info/stream_info.h#L419

Should we make that `absl::optional<uint64_t> bytesSent()` to centralize the process of determining if there was a body or not?

Ditto for `bytesReceived()`:

https://github.com/envoyproxy/envoy/blob/main/envoy/stream_info/stream_info.h#L296

Reply:
Ok I added accounting for 0 length bodies and empty bodies back. We can change it or add a knob to configure this later on, if there's a need for excluding them.